Defer transaction signing until user clicks Send

151henry151 opened the pull request:
Fixes [#30070](https://github.com/bitcoin/bitcoin/issues/30070)

When creating an unsigned PSBT from the GUI, the transaction was already signed during preparation, causing legacy inputs to have non-empty scriptSig fields. The PSBT parser then rejects them.

This defers signing until the user clicks "Send" instead of signing during preparation. Fee calculation still works since transactions can be created without signing.

Follows the approach suggested by @achow101 in the issue comments.

151henry151 committed:
qt: Defer transaction signing until user clicks Send

This fixes issue #30070 where creating unsigned PSBTs from the GUI
would fail because the transaction was already signed during preparation,
causing legacy inputs to have non-empty scriptSig fields.

The fix defers signing until the user explicitly clicks 'Send', allowing
truly unsigned PSBTs to be created while still supporting fee calculation.

151henry151 commented:
It looks like the one failing check is a CI issue with disk space too low. I'm not sure how to address that. I think the code of the PR is correct.

151henry151 commented:
@achow101 would you like to check this out and let me know if it looks OK?

151henry151 commented:
Should I push an empty commit to trigger a re-run of the CI here, or is there some other action needed?

achow101 commented on src/qt/walletmodel.cpp:
```diff
@@ -147,7 +147,7 @@ bool WalletModel::validateAddress(const QString& address) const
 }
 
-WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransaction &transaction, const CCoinControl& coinControl)
+WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransaction &transaction, const CCoinControl& coinControl, bool sign)
```
`prepareTransaction` has only a single caller that already sets `sign = false`. I don't think this needs to take a parameter, it can just always not sign.

151henry151 replied:
suggestion implemented

achow101 reviewed (commented):
Concept ACK

151henry151 commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -540,6 +542,24 @@ void SendCoinsDialog::sendButtonClicked([[maybe_unused]] bool checked)
+            PartiallySignedTransaction psbtx(mtx);
+            bool complete = false;
+            // Fill and sign the PSBT
+            const auto err{model->wallet().fillPSBT(std::nullopt, /*sign=*/true, /*bip32derivs=*/false, /*n_signed=*/nullptr, psbtx, complete)};
```
refactored as suggested

johnny9 reviewed (commented):
Double check this works with locked encrypted wallets

johnny9 reviewed (changes requested):
Need to move the UnlockContext when you sign the transaction otherwise it fails to send now.

https://github.com/user-attachments/assets/6b7e6894-b803-4753-9f20-66541af63633

151henry151 commented on src/qt/sendcoinsdialog.cpp:
@johnny9 I think I've addressed this correctly with [39a5fed](https://github.com/bitcoin-core/gui/commit/39a5fed760273d796d6b7e4cb298fc36d4d27022) -- it passed my local tests; do you want to check it out and see if it looks good to you?

johnny9 commented on src/qt/sendcoinsdialog.cpp:
This is a little different now that the current transaction isn't signed yet so the kvB is different now than before.  Can either change the dialog to make this clear or adjust this size to be the same as the estimated size the fee calculation uses.

johnny9 reviewed (commented):
Moving the UnlockContext in the "Send" click flow resolved the previous issue. I think the previous UnlockContext is no longer needed.

I also see that the send confirmation dialog isn't as accurate as it was previously so should adjust what the dialog says or expose the correct fee estimated size through the model.

johnny9 commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -269,23 +272,18 @@ bool SendCoinsDialog::PrepareSendText(QString& question_string, QString& informa
-    WalletModel::UnlockContext ctx(model->requestUnlock());
-    if(!ctx.isValid())
-    {
-        // Unlock wallet was cancelled
```
I think I was wrong about removing this. I think you will need unlock here too to be sure you can get a change address.

johnny9 reviewed (commented):
I actually think I was wrong and we need both UnlockContexts. Sorry to flp-flop on that.  Also take a look at achow's comment again. Seems like you didnt remove the parameter he mentioned. Other than those things I think this looks good now

151henry151 commented:
@johnny9 @achow101 I think this is ready for review, if anybody wants to check it out?

151henry151 committed:
qt: Defer transaction signing until user clicks Send

This fixes issue #30070 where creating unsigned PSBTs from the GUI
would fail because the transaction was already signed during preparation,
causing legacy inputs to have non-empty scriptSig fields.

The fix defers signing until the user explicitly clicks 'Send', allowing
truly unsigned PSBTs to be created while still supporting fee calculation.
